### PR TITLE
Bug 1712080 - Gzip ping paylod before upload

### PR DESCRIPTION
**Pull request by brizental**
Should I file a bug to document that pings are gzipped? I looked and that is not mentioned on the Glean book. That could be important for folks using custom uploaders for testing, but that could also be behaviour we want to disencourage... Even if we don't disencourage it, we might want to expose the [WaitableHttpClient](https://github.com/mozilla/glean.js/blob/main/glean/tests/integration/schema/schema.spec.ts#L46-L81) in the future. Then people still don't need to worry about gzipping.

### Pull Request checklist ###

- [x] **Quality**: Make sure this PR builds and runs cleanly.
  - Inside the `glean/folder`, run:
    - `npm run test` Runs _all_ tests
    - `npm run lint && npm run lint:circular-deps` Runs _all_ linters
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry to `CHANGELOG.md` or an explanation of why it does not need one
- [ ] **Documentation**: This PR includes documentation changes, an explanation of why it does not need that or a follow-up bug has been filed to do that work
  - Documentation should be added to [The Glean Book](https://mozilla.github.io/glean/book/index.html) when making changes to Glean's user facing API
    - When changes to the Glean Book are necessary, link to the corresponding PR on the [`mozilla/glean`](https://github.com/mozilla/glean) repository
  - Documentation should be added to [the Glean.js developer documentation](https://github.com/mozilla/glean.js/tree/main/docs) when making internal code changes

### CODE REVIEW

File: glean/src/core/upload/index.ts [hunk 1]
```diff
@@ -124,25 +126,35 @@ class PingUploader implements PingsDatabaseObserver {
    */
   private async preparePingForUpload(ping: QueuedPing): Promise<{
     headers: Record<string, string>,
-    payload: string
+    payload: string | Uint8Array
```
Comment (Dexterp37) [inline glean/src/core/upload/index.ts:129]: Would you kindly document why accepting both is needed? (QML not supporting encoding easily?)

File: glean/src/platform/qt/fflate.stub.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+/**
+ * This file is a drop in replacement for `fflate` in QML.
+ * This library raises errors in QML. https://github.com/101arrowz/fflate/issues/71
+ * could be the the culprit, but it is unclear at this point.
+ */
+
+// eslint-disable-next-line jsdoc/require-jsdoc
+export function gzipSync() {
+  throw new Error("Attempted to use `gzipSync` from QML, but that is not supported.");
```
Comment (Dexterp37) [inline glean/src/platform/qt/fflate.stub.js:13]: Can you add a comment explaining why we throw and a link to the bug for implementing this, if we're planning to

File: glean/src/core/upload/index.ts (hunk 1 above)
Comment (brizental) [reply glean/src/core/upload/index.ts:129]: Well, WML support is not the only reason. We do the same in glean-core, gzipping action may fail and it it does then we send the unzipped ping. I will add a line about this though :)

**Review by Dexterp37 — APPROVED** [review]
Would you kindly do, if you didn't already, a test with the debug view (before merging)?

**Comment by brizental** [comment]
> Would you kindly do, if you didn't already, a test with the debug view (before merging)?

https://debug-ping-preview.firebaseapp.com/pings/gzipped-webext Did that while developing, ignore the failing one that was just _development_.

**Pull request merged**